File: 백준/Gold/3079. 입국심사/입국심사.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
time = sorted(list(int(input()) for _ in range(N)))

start = 0
# time의 최소값 * 사람 수 => 최대 시간 
end = min(time) * M
result = 0

while start <= end :
    mid = (start + end) // 2
    quotient = 0
    for t in time:
        quotient += mid // t

    if quotient >= M:
        end = mid - 1
        result = mid
    else:
        start = mid + 1

print(result)

File: 백준/Gold/3079. 입국심사/입국심사.js
const solution = (M, waitlist) => {
  waitlist.sort((a, b) => b - a);
  let result = waitlist[0] * M;
  let [start, end] = [0, result];
  let mid;

  while (start < end) {
    mid = Math.floor((start + end) / 2);
    const sum = waitlist.reduce((prev, cur) => prev + Math.floor(mid / cur), 0);

    if (mid === end) {
      break;
    }
    if (sum < M) {
      start = mid + 1;
    } else {
      result = mid;
      end = mid;
    }
  }

  console.log(result);
};

const read = () => {
  let input = require("fs")
    .readFileSync("dev/stdin")
    .toString()
    .trim()
    .split("\n");
  const [NM, ...waitlist] = input;
  const M = Number.parseInt(NM.split(" ").pop(), 10);

  solution(
    M,
    waitlist.map((x) => Number.parseInt(x, 10))
  );
};

read();
